### PR TITLE
Improve packed_transaction parsing performance

### LINKED ISSUE #1025: Improve packed_transaction parsing performance
See https://github.com/AntelopeIO/leap/pull/999#discussion_r1165459091
Simple speedup is to create a caching resolver that caches abis as it goes instead of repeatedly retrieving them.

### PR DESCRIPTION
**Pull request by greg7mdp**
Resolves #1025.

Introduce a `caching_resolver `class, which provides the same API as a resolver, but which caches `abi_serializers`  as they are retrieved, so we pay the cost of retrieval only once.

### CODE REVIEW

File: libraries/chain/include/eosio/chain/abi_serializer.hpp [hunk 1]
```diff
@@ -1066,8 +1068,39 @@ class abi_serializer_cache_builder {
...
+   std::optional<std::reference_wrapper<const abi_serializer>> operator()(const account_name& account) const {
+      auto it = abi_serializers.find(account);
+      if (it != abi_serializers.end() && it->second)
+         return *it->second;
+      try {
+         auto& dest = abi_serializers[account];
+         dest = resolver_(account);
+         return *dest;
+      } catch( ... ) {
+         throw; // throw if embedded resolver throws
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/abi_serializer.hpp:1096]: Why catch and throw here?

File: libraries/chain/include/eosio/chain/abi_serializer.hpp [hunk 2]
```diff
@@ -825,13 +825,14 @@ namespace impl {
                from_variant(data, act.data);
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/abi_serializer.hpp:825]: The `extract` on line 739 takes `Resolver` by value. I would hope the compiler would remove that copy, but best to change it to a `const&` I think.

**Review by heifner — CHANGES_REQUESTED** [review]

**Comment by heifner** [comment]
> @heifner if I create a `abi_serializer` cache when parsing the `packed_transaction` sent by the user, can I reuse the same cache for serializing the trace result? In my current version we do recreate the cache from the trace.

Short answer:  no. 
The transaction itself or another transaction executed between the time of trx input and trx execution can `setabi` which would change the abi. Your cache could check for a diff `abi_sequence` on the account, but you would have to look up the account.  You would save the creation of the `abi_serializer`, but not sure it is worth the extra complexity.

File: libraries/chain/include/eosio/chain/abi_serializer.hpp (hunk 2 above)
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/abi_serializer.hpp:825]: yes for sure I'll change it.!

File: libraries/chain/include/eosio/chain/abi_serializer.hpp [hunk 3]
```diff
@@ -1066,8 +1068,39 @@ class abi_serializer_cache_builder {
...
+   resolver_fn_t resolver_;
    abi_serializer_cache_t abi_serializers;
 };
 
+/*
+ * This is equivalent to a resolver, except that everytime the abi_serializer for an account 
+ * is retrieved, it is stored in an unordered_map, so we won't waste time retrieving it again.
+ * This is handy when parsing packed_transactions received in a fc::variant.
+ */
+class caching_resolver {
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/abi_serializer.hpp:1080]: sure.

File: libraries/chain/include/eosio/chain/abi_serializer.hpp (hunk 2 above)
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/abi_serializer.hpp:825]: done

**Comment by greg7mdp** [comment]
Looks like I have some test failures. I looked at one an it looks like we are trying to retrieve the account before we create it? Am I understanding this right?

image

**Comment by heifner** [comment]
> Looks like I have some test failures. I looked at one an it looks like we are trying to retrieve the account before we create it? Am I understanding this right?

You can ignore that. It is expected.

**Comment by heifner** [comment]
> > Looks like I have some test failures. I looked at one an it looks like we are trying to retrieve the account before we create it? Am I understanding this right?
> 
> You can ignore that. It is expected.

See telegram.

**Comment by greg7mdp** [comment]
Hey, thanks for the tip, I debugged with the core file (super easy)  and it pointed me to the problem. Apparently the cache was returning an incorrect abi_serializer. I'm not 100% sure why, but I think it was because we construct a `std::optional` containing a `std::reference_wrapper`, and when constructing it from an empty `std::optional<abi_serializer>` it didn't work quite as expected. Anyways I tried to be very explicit and it seems to work.

File: libraries/chain/include/eosio/chain/abi_serializer.hpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/include/eosio/chain/abi_serializer.hpp:1108]: The catch-throw is still here.

Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/abi_serializer.hpp]: Excellent point, it is unneeded. Removed it!

**Review by heifner — APPROVED** [review]

**Review by 766C6164 — APPROVED** [review]

**Pull request merged**